python/private/rule_builders.bzl

```diff
@@ -0,0 +1,673 @@
+:::
+
+Example usage:
+
+```
+
+load(":rule_builders.bzl", "ruleb")
+load(":attr_builders.bzl", "attrb")
+
+# File: foo_binary.bzl
+_COMMON_ATTRS = {
+    "srcs": lambda: attrb.LabelList(...),
+}
+
+def create_foo_binary_builder():
+    foo = ruleb.Rule(
+        executable = True,
+    )
+    foo.implementation.set(_foo_binary_impl)
+    foo.attrs.update(COMMON_ATTRS)
+    return foo
+
+def create_foo_test_builder():
+    foo = create_foo_binary_build()
+
+    binary_impl = foo.implementation.get()
+    def foo_test_impl(ctx):
+      binary_impl(ctx)
+      ...
+
+    foo.implementation.set(foo_test_impl)
+    foo.executable.set(False)
+    foo.test.test(True)
+    foo.attrs.update(
+        _coverage = attrb.Label(default="//:coverage")
+    )
+    return foo
+
+foo_binary = create_foo_binary_builder().build()
+foo_test = create_foo_test_builder().build()
+
+# File: custom_foo_binary.bzl
+load(":foo_binary.bzl", "create_foo_binary_builder")
+
+def create_custom_foo_binary():
+    r = create_foo_binary_builder()
+    r.attrs["srcs"].default.append("whatever.txt")
+    return r.build()
+
+custom_foo_binary = create_custom_foo_binary()
+```
+"""
+
+load("@bazel_skylib//lib:types.bzl", "types")
+load(
+    ":builders_util.bzl",
+    "kwargs_getter",
+    "kwargs_set_default_dict",
+    "kwargs_set_default_doc",
+    "kwargs_set_default_ignore_none",
+    "kwargs_set_default_list",
+    "kwargs_setter",
+    "list_add_unique",
+)
+
+def _ExecGroup_typedef():
+    """Builder for {external:bzl:obj}`exec_group`
+
+    :::{function} toolchains() -> list[ToolchainType]
+    :::
+
+    :::{function} exec_compatible_with() -> list[str | Label]
+    :::
+
+    :::{include} /_includes/field_kwargs_doc.md
+    :::
+    """
+
+def _ExecGroup_new(**kwargs):
+    """Creates a builder for {external:bzl:obj}`exec_group`.
+
+    Args:
+        **kwargs: Same as {external:bzl:obj}`exec_group`
+
+    Returns:
+        {type}`ExecGroup`
+    """
+    kwargs_set_default_list(kwargs, "toolchains")
+    kwargs_set_default_list(kwargs, "exec_compatible_with")
+
+    for i, value in enumerate(kwargs["toolchains"]):
+        kwargs["toolchains"][i] = _ToolchainType_maybe_from(value)
+
+    # buildifier: disable=uninitialized
+    self = struct(
+        toolchains = kwargs_getter(kwargs, "toolchains"),
+        exec_compatible_with = kwargs_getter(kwargs, "exec_compatible_with"),
+        kwargs = kwargs,
+        build = lambda: _ExecGroup_build(self),
+    )
+    return self
+
+def _ExecGroup_maybe_from(obj):
+    if types.is_function(obj):
+        return obj()
+    else:
+        return obj
+
+def _ExecGroup_build(self):
+    kwargs = dict(self.kwargs)
+    if kwargs.get("toolchains"):
+        kwargs["toolchains"] = [
+            v.build() if hasattr(v, "build") else v
+            for v in kwargs["toolchains"]
+        ]
+    if kwargs.get("exec_compatible_with"):
+        kwargs["exec_compatible_with"] = [
```

Comment:
This file has similar comments to the `attr_builders` - mainly string constants?

Reply:
Will do, mostly. I'm not a huge fan of creating symbolic constants for _every_ string literal because it can harm readability, but you have a good point -- many of these strings are repeated many times across many functions, sometimes in "optional" contexts (dict.get, key in dict) where typos would be easy to miss. I rewrote a few to use constants and it is still pretty readable. I'll update literals that span multiple functions.

Reply:
Done